IdentityServer/Core/ResourceOwnerPasswordValidator.cs

```diff
@@ -0,0 +1,59 @@
+using System.Collections.Generic;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using IdentityModel;
+using IdentityServer.Core.Users;
+using IdentityServer4.Models;
+using IdentityServer4.Validation;
+
+namespace IdentityServer.Core
+{
+    public class ResourceOwnerPasswordValidator : IResourceOwnerPasswordValidator
+    {
+        private readonly IUserIdentityService _userIdentityService;
+
+
+        public ResourceOwnerPasswordValidator(IUserIdentityService userIdentityService)
+        {
+            _userIdentityService = userIdentityService;
+        }
+
+        public async Task ValidateAsync(ResourceOwnerPasswordValidationContext context)
+        {
+            try
+            {
+                var userName = context.UserName;
+                var user = await _userIdentityService.FindByNameAsync(userName);
+                if (user != null)
+                {
+                    if (await _userIdentityService.CheckPasswordAsync(user, context.Password))
+                    {
+                        context.Result = new GrantValidationResult(
+                            subject: user.Id.ToString(),
+                            authenticationMethod: "password",
+                            claims: GetUserClaims(user));
+
+                        return;
+                    }
+
+                    context.Result = new GrantValidationResult(TokenRequestErrors.InvalidGrant, "Incorrect password");
+                    return;
+                }
+                context.Result = new GrantValidationResult(TokenRequestErrors.InvalidGrant, "User does not exist.");
```

Comment:
Hmm. Do we really wanna tell him that he typed ONLY wrong password (for example haker is looking for existing emails to send him some spam)? Less info before auth for me it's better. What do you think? I mean it's good but I'm not sure because our frontend will probably check only status which will be the same for every responses(I don't think that checking response message it's good idea).

Reply:
Nice point, i'll change the message to something less specific.